cupy/_core/_gufuncs.py

```diff
@@ -387,12 +316,98 @@ def __init__(self, func, signature, **kwargs):
+        # https://github.com/dask/dask/blob/61b578f5a3ad88cbc6a8b9a73ce08c551bd969fa/dask/array/gufunc.py#L58-L172
+        if axes is not None and axis is not None:
+            raise ValueError(
+                'Only one of `axis` or `axes` keyword arguments '
+                'should be given')
+        if axes and not isinstance(axes, list):
+            raise ValueError('`axes` has to be of type list')
+
+        if keepdims:
+            # Store the number of core dimensions to append into keepdims.
+            keepdims = None
+            for icd in self._input_coredimss:
+                ndim_core = len(icd)
+                if keepdims is not None and keepdims != ndim_core:
+                    raise ValueError(
+                        '`keepdims` requires all inputs to have the same '
+                        'number of core dimensions')
+                keepdims = ndim_core
+
+            for ocd in self._output_coredimss:
+                if len(ocd) > 0:
+                    raise ValueError(
+                        '`keepdims` can only be used for scalar outputs')
+
+        if axis is not None:
+            if not isinstance(axis, int):
+                raise ValueError('`axis` argument has to be an integer value')
+            if self._core_num_dim_ix != 1:
+                raise ValueError(
+                    '`axis` can be used only, if there is a single shared '
+                    f'core dimension, but {self._core_num_dim_ix} are '
+                    f'present in the signature {self._signature}'
+                )
+            if not keepdims:
+                axes = [(axis,) if cd else () for cd in self._core_dims]
+            else:
+                # When keepdims is used with axis, it applies to the output
+                axes = [(axis,) if cd else () for cd in self._input_coredimss]
+                axes.extend([(axis,)] * self._nout)
+        elif axes is not None:
+            # Allow a single integer in axes.
+            axes = [(a,) if isinstance(a, int) else a for a in axes]
+        else:
+            axes = [None] * (self._nin + self._nout)
+
+        if len(axes) == self._nin:
+            for ocd in self._output_coredimss:
+                if len(ocd) > 0:
+                    raise ValueError(
+                        '`axes` entries for outputs can only be omitted if '
+                        'none of them has core axes.')
+            # Output axes may be ommitted, fill them in if this is the case.
+            # This seems actually more relaxes as NumPy as of NumPy 2.4.
+            axes.extend([None] * self._nout)
+        elif len(axes) != self._nin + self._nout:
+            raise ValueError(
+                'The number of `axes` entries is not equal to the number of '
+                'input and output arguments. Outputs may only be omitted if '
+                'they are all scalar.')
+
+        return axes[:self._nin], axes[self._nin:], keepdims
```

Comment:
This one is actually a bit similar, but moved wholesale.  The big difference is to not fill to fill in `None` as axes when neither is passed.
That information is useful later and also allows side-stepping the transposes a bit more easily.